### PR TITLE
STM - MQTT Functionality

### PR DESCRIPTION
**Pull request by DylanCavers**
Adding MQTT functionality to the state machine in order to receive transitions and publish the current state

### CODE REVIEW

File: test/state_machine/stm_tests.cpp [hunk 1]
```diff
@@ -12,9 +15,26 @@ void testTransition(std::unique_ptr<state_machine::StateMachine> &stm,
   ASSERT_TRUE(stm->getCurrentState() == expected_state);
 }
 
+std::shared_ptr<core::Mqtt> getMqtt()
```
Comment (TomLonergan03) [inline test/state_machine/stm_tests.cpp:18]: unless you have a way to use the broker as part of the tests, you're gonna want an `IMqtt` interface and a `MockMqtt` class in `lib/utils` I reckon

File: lib/state_machine/state_machine.cpp [hunk 2]
```diff
@@ -35,4 +41,42 @@ core::Result StateMachine::handleTransition(const State &state)
...
+  rapidjson::Value state_value(state_string.c_str(), message_payload->GetAllocator());
+  message_payload->AddMember("transition", state_value, message_payload->GetAllocator());
+  core::MqttMessage::Header header{.timestamp = 0, .priority = core::MqttMessagePriority::kNormal};
+  const core::MqttMessage message{topic, header, message_payload};
+  mqtt_->publish(message, core::MqttMessageQos::kExactlyOnce);
+}
+
+void StateMachine::startStateMachine()
+{
+  while (true) {
```
[the remark this answers is not in the document]
Comment (DylanCavers) [reply lib/state_machine/state_machine.cpp]: updated to finish when a shutdown command is received

File: lib/utils/dummy_mqtt.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{
+  return core::Result::kSuccess;
+}
+
+core::Result MockMqtt::consume()
+{
+  return core::Result::kSuccess;
+}
+
+std::optional<core::MqttMessage> MockMqtt::getMessage()
```
[the remark this answers is not in the document]
Comment (DylanCavers) [reply lib/utils/dummy_mqtt.cpp:27]: will probably leave this till later when we think about testing

[the remark this answers is not in the document]
Comment (TomLonergan03) [reply lib/utils/dummy_mqtt.cpp:27]: Fair enough

File: lib/state_machine/state_machine.cpp [hunk 4]
```diff
@@ -35,4 +42,43 @@ core::Result StateMachine::handleTransition(const State &state)
...
+  mqtt_->publish(message, core::MqttMessageQos::kExactlyOnce);
+}
+
+void StateMachine::startStateMachine()
+{
+  while (StateMachine::getCurrentState() != State::kShutdown) {
+    auto updated = StateMachine::update();
+    StateMachine::publishCurrentState();
+  }
+  StateMachine::publishCurrentState();
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:81]: these aren't static methods

File: lib/state_machine/state_machine.cpp [hunk 5]
```diff
@@ -35,4 +42,43 @@ core::Result StateMachine::handleTransition(const State &state)
...
+  message_payload->AddMember("transition", state_value, message_payload->GetAllocator());
+  core::MqttMessage::Header header{.timestamp = 0, .priority = core::MqttMessagePriority::kNormal};
+  const core::MqttMessage message{topic, header, message_payload};
+  mqtt_->publish(message, core::MqttMessageQos::kExactlyOnce);
+}
+
+void StateMachine::startStateMachine()
+{
+  while (StateMachine::getCurrentState() != State::kShutdown) {
+    auto updated = StateMachine::update();
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:78]: ??

if the return value matters we should do something with it, if not get rid of the return

File: lib/state_machine/transition_table.hpp [hunk 6]
```diff
@@ -37,7 +37,8 @@ const TransitionTable transition_to_state_dynamic
      {{State::kBatteryRecharge, State::kFailure}, State::kFailure},
      {{State::kFailureBrake, State::kCapacitorDischarge}, State::kCapacitorDischarge},
      {{State::kFailure, State::kCapacitorDischarge}, State::kCapacitorDischarge},
-     {{State::kCapacitorDischarge, State::kSafe}, State::kSafe}};
+     {{State::kCapacitorDischarge, State::kSafe}, State::kSafe},
+     {{State::kSafe, State::kShutdown}, State::kShutdown}};
```
Comment (TomLonergan03) [inline lib/state_machine/transition_table.hpp:41]: We probably want to shut down from some (I'm not sure which) pre-run states

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.cpp (hunk 4 above)
Comment (DylanCavers) [reply lib/state_machine/state_machine.cpp:81]: do we want them as static?

Comment (TomLonergan03) [reply lib/state_machine/state_machine.cpp]: i mean you should construct a state machine and do `state_machine.update()`. I don't think these can be made static but they definitely should not be invoked as if they are

Comment (DylanCavers) [reply lib/state_machine/state_machine.cpp]: nvm this was just a skill issue from me

File: lib/state_machine/state_machine.cpp [hunk 7]
```diff
@@ -35,4 +42,38 @@ core::Result StateMachine::handleTransition(const State &state)
...
+  std::shared_ptr message_payload = std::make_shared<rapidjson::Document>();
+  message_payload->SetObject();
+  rapidjson::Value state_value(state_string.c_str(), message_payload->GetAllocator());
+  message_payload->AddMember("transition", state_value, message_payload->GetAllocator());
+  core::MqttMessage::Header header{.timestamp = 0, .priority = core::MqttMessagePriority::kNormal};
+  const core::MqttMessage message{topic, header, message_payload};
+  mqtt_->publish(message, core::MqttMessageQos::kExactlyOnce);
+}
+
+void StateMachine::startStateMachine()
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:70]: `run()` imo

File: lib/state_machine/state_machine.cpp [hunk 8]
```diff
@@ -35,4 +42,38 @@ core::Result StateMachine::handleTransition(const State &state)
   return core::Result::kFailure;
 }
 
+void StateMachine::update()
+{
+  mqtt_->consume();
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:47]: is it cleaner to call this here or in `StateMachine::run()`? idk

File: lib/state_machine/state_machine.cpp [hunk 9]
```diff
@@ -35,4 +42,38 @@ core::Result StateMachine::handleTransition(const State &state)
...
+
+void StateMachine::publishCurrentState()
+{
+  const auto state_string         = stateToString(getCurrentState());
+  const auto topic                = core::MqttTopic::kTest;
+  std::shared_ptr message_payload = std::make_shared<rapidjson::Document>();
+  message_payload->SetObject();
+  rapidjson::Value state_value(state_string.c_str(), message_payload->GetAllocator());
+  message_payload->AddMember("transition", state_value, message_payload->GetAllocator());
+  core::MqttMessage::Header header{.timestamp = 0, .priority = core::MqttMessagePriority::kNormal};
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:65]: `const`

File: lib/state_machine/state_machine.hpp [hunk 10]
```diff
@@ -8,22 +8,27 @@
 #include <queue>
 #include <unordered_map>
 
+#include "core/mqtt.hpp"
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:11]: <angle brackets>

File: lib/state_machine/state_machine.hpp [hunk 11]
```diff
@@ -58,6 +64,7 @@ class StateMachine {
        {State::kSafe, "kSafe"}};
 
   State current_state_;
+  std::shared_ptr<core::IMqtt> mqtt_;
   TransitionTable transition_to_state_;
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:68]: `const`

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.cpp [hunk 12]
```diff
@@ -35,4 +42,39 @@ core::Result StateMachine::handleTransition(const State &state)
...
+  const core::MqttMessage::Header header{.timestamp = 0,
+                                         .priority  = core::MqttMessagePriority::kNormal};
+  const core::MqttMessage message{topic, header, message_payload};
+  mqtt_->publish(message, core::MqttMessageQos::kExactlyOnce);
+}
+
+void StateMachine::run()
+{
+  while (getCurrentState() != State::kShutdown) {
+    mqtt_->consume();
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:73]: this returns `kFailure` if it receives an invalid message, we should prob enter shutdown state in that case imo

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]
not much left

File: lib/state_machine/state_machine.cpp (hunk 12 above)
Comment (DylanCavers) [reply lib/state_machine/state_machine.cpp:73]: atm moment we can only transition to `kshutdown` from `kSafe`, should I add a transition from other states?

**Review by TomLonergan03 — APPROVED** [review]

**Review by lsalhani — APPROVED** [review]

**Pull request merged**